ai/src/inference.py

```diff
@@ -0,0 +1,135 @@
+                response=aiengine_pb2.Response(
+                    result=result, message=message, error=True
+                )
+            )
+
+        return None
+
+    def __load_agent(self, model_exists) -> SpiceAIAgent:
+        # Ideally we could just re-use the in-memory agent we created during training,
+        # but tensorflow has issues with multi-threading in python, so we are just loading it from the file system
+        data_manager = self.data_managers[self.request.pod]
+        model_data_shape = data_manager.get_shape()
+        if model_exists:
+            save_path = Trainer.SAVED_MODELS[self.request.pod]
+            with open(save_path / "meta.json", "r", encoding="utf-8") as meta_file:
+                save_info = json.loads(meta_file.read())
+            algorithm = save_info["algorithm"]
+        else:
+            algorithm = "dql"
+
+        agent: SpiceAIAgent = get_agent(algorithm, model_data_shape, len(data_manager.action_names))
+        if model_exists:
+            agent.load(Path(save_path))
+        return agent
+
+    def get_result(self):
+        try:
+            model_exists = False
+            if self.request.pod in Trainer.SAVED_MODELS:
+                model_exists = True
+
+            error = self.__validate_request()
+            if error is not None:
+                return error
+
+            agent = self.__load_agent(model_exists=model_exists)
+
+            data_manager = self.data_managers[self.request.pod]
+
+            if data_manager.massive_table_filled.shape[0] < data_manager.get_window_span():
+                return aiengine_pb2.InferenceResult(
+                    response=aiengine_pb2.Response(result="not_enough_data", error=True))
+
+            if self.use_latest_time:
+                latest_window = data_manager.get_latest_window()
+                state = data_manager.flatten_and_normalize_window(latest_window)
+            else:
+                requested_window = data_manager.get_window_at(self.inference_time)
+                state = data_manager.flatten_and_normalize_window(requested_window)
+
+            action_from_model, probabilities = agent.act(state)
+        except InvalidDataShapeException as ex:
+            result = ex.type
+            message = ex.message
+
+            if Trainer.TRAINING_LOCK.locked():
+                result = "training_not_complete"
+                message = (
+                    "Please wait to get a recommendation until after training completes"
+                )
+
+            return aiengine_pb2.InferenceResult(
+                response=aiengine_pb2.Response(
+                    result=result, message=message, error=True
+                )
+            )
+        confidence = f"{probabilities[action_from_model]:.3f}"
+        if not model_exists:
+            confidence = 0.0
+
+        action_name = data_manager.action_names[action_from_model]
+
+        if self.use_latest_time:
```

Comment:
I like ternary for this kind of blocks as it is 1 line instead of 4 and faster execution in python. But this is really an opinion, feel free to ignore if you think branches are better.

Reply:
If I make this a ternary it is over 120 characters long in a single line. I'll leave it as is for now, but I'll consider it going forward!